node-graph/gcore/src/graphic_element/renderer.rs

```diff
@@ -0,0 +1,188 @@
+			transform: DAffine2::IDENTITY,
+			image_data: Vec::new(),
+		}
+	}
+
+	/// Add an outer `<svg />` tag with a `viewBox` and the `<defs />`
+	pub fn format_svg(&mut self, bounds_min: DVec2, bounds_max: DVec2) {
+		let (x, y) = bounds_min.into();
+		let (size_x, size_y) = (bounds_max - bounds_min).into();
+		let defs = &self.svg_defs;
+		let svg_header = format!(r#"<svg xmlns="http://www.w3.org/2000/svg" viewBox="{x} {y} {size_x} {size_y}"><defs>{defs}</defs>"#,);
+		self.svg.insert(0, svg_header.into());
+		self.svg.push("</svg>".into());
+	}
+}
+
+/// Static state used whilst rendering
+pub struct RenderParams {
+	pub view_mode: crate::vector::style::ViewMode,
+	pub culling_bounds: Option<[DVec2; 2]>,
+	pub thumbnail: bool,
+}
+
+impl RenderParams {
+	pub fn new(view_mode: crate::vector::style::ViewMode, culling_bounds: Option<[DVec2; 2]>, thumbnail: bool) -> Self {
+		Self { view_mode, culling_bounds, thumbnail }
+	}
+}
+
+fn format_transform_matrix(transform: DAffine2) -> String {
+	transform.to_cols_array().iter().map(ToString::to_string).collect::<Vec<_>>().join(",")
+}
+
+pub trait GraphicElementRendered {
+	fn render_svg(&self, render: &mut SvgRender, render_params: &RenderParams);
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]>;
+}
+
+impl GraphicElementRendered for GraphicGroup {
+	fn render_svg(&self, render: &mut SvgRender, render_params: &RenderParams) {
+		self.iter().for_each(|element| element.graphic_element_data.render_svg(render, render_params))
+	}
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]> {
+		self.iter().filter_map(|element| element.graphic_element_data.bounding_box(transform)).reduce(combine_bounds)
+	}
+}
+
+impl GraphicElementRendered for VectorData {
+	fn render_svg(&self, render: &mut SvgRender, render_params: &RenderParams) {
+		let layer_bounds = self.bounding_box().unwrap_or_default();
+		let transfomed_bounds = self.bounding_box_with_transform(render.transform).unwrap_or_default();
+
+		render.svg.push("<path d=\"".into());
+		let mut path = String::new();
+		for subpath in &self.subpaths {
+			let _ = subpath.subpath_to_svg(&mut path, self.transform * render.transform);
+		}
+		render.svg.push(path.into());
+		render.svg.push("\"".into());
+
+		let style = self.style.render(render_params.view_mode, &mut render.svg_defs, render.transform, layer_bounds, transfomed_bounds);
+		render.svg.push(style.into());
+		render.svg.push("/>".into());
+	}
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]> {
+		self.bounding_box_with_transform(self.transform * transform)
+	}
+}
+
+impl GraphicElementRendered for Artboard {
+	fn render_svg(&self, render: &mut SvgRender, render_params: &RenderParams) {
+		self.graphic_group.render_svg(render, render_params)
+	}
+	fn bounding_box(&self, transform: DAffine2) -> Option<[DVec2; 2]> {
+		let artboard_bounds = self.bounds.map(|[a, b]| (transform * Quad::from_box([a.as_dvec2(), b.as_dvec2()])).bounding_box());
```

Comment:
Isn't that just the same as:
```rust
let artboard_bounds = self.bounds.map(|points| points.map(|c| transform.transform_point2(c.as_dvec2())))
```

Reply:
No, it isn't. Note that when rotating a rectangle, the axis aligned bounding box is no longer simply the transformed positions of the top right and bottom left corner:
![image](https://github.com/GraphiteEditor/Graphite/assets/78500760/4c2ac97d-93ff-45ca-aa4c-f7b9fca2efdd)

Reply:
oh right, somehow my brain ignored the existence of rotation.